src/mailadm/web.py

```diff
@@ -25,12 +26,13 @@ def new_email():
             token_info = conn.get_tokeninfo_by_token(token)
             if token_info is None:
                 return "token {} is invalid".format(token), 403
-
-            try:
-                user_info = conn.add_email_account(token_info, tries=10)
-                conn.gen_sysfiles()
-            except DBError as e:
-                return str(e), 409
-            return jsonify(email=user_info.addr, password=user_info.clear_pw,
-                           expiry=token_info.expiry, ttl=user_info.ttl)
+            tries = 10
+            for i in range(tries):
+                try:
+                    user_info = conn.add_email_account(token_info)
```

Comment:
any particular reason why you removed the "tries" parameter to add_email_account?

Reply:
previously, add_email_account() wasn't holding more than the retry logic. I always found it odd that the user creation logic was split over 3 different methods in conn.py alone (_add_addr, add_user, and add_email_account); so I moved that logic out of conn.py.

Other than the web interface the retrying doesn't make much sense, I would expect the cmdline or bot interface to just throw an error when it doesn't work immediately. We just don't do it in the web interface, because that's usually used by users, not by admins, and an error would be useless to them; so it makes sense to just let web.py retry it a few times. I thought that it would be better to handle web-exclusive logic in web.py.

Reply:
not sure it's just about the web.  We are creating a random string out of five characters for each random account.  If you roll the five-char dice 10K times chances are (if i am not mistaken, see e.g. https://blogs.sas.com/content/iml/2013/07/01/duplicates-in-tossing-dice.html ) 78% that you get a duplicate. Logically, it's ok to ask the db to generate a unique new account and let it work a bit to figure it out.

Reply:
Okay, I re-introduced it - but I don't want to use the old method names, they really confused me when I tried to get into the code. The whole scheme that there were three different methods in conn.py to add users has confusion potential. The old scheme was:
- add_email_account: retry logic, calls _add_addr
- _add_addr: generated addr & password, calls add_user
- add_user: added the user to the database, implying that the sysfiles would be generated.

Now we could do two different things:

Alternative 1 (this branch right now):

- add_email_account_tries: retry logic calls add_email_account
- add_email_account: generates addr & password, calls mailcow.add_user, calls add_user_db
- add_user_db: adds user to the database

Alternative 2 would be a bit different. It assumes that we don't need the retry logic for anything else than generating addresses (is there anything else?)

- add_email_account: calls generate_addr, calls mailcow.add_user, calls add_user_db
- generate_addr: tries to generate an addr for $tries times, and asks mailcow if the user exists already
- add_user_db: adds the user to the mailadm.db

And Alternative 3 is something I'm planning to do with the bot refactoring anyway:

- In Alt3 there are 3 user interfaces: bot.py, cmdline.py, and web.py.
- they all call commands.py:add_user, which calls conn.generate_addr, conn.add_user, and mailcow.add_user.
- conn.generate_addr: tries to generate an address for $tries times, and asks mailcow if the user exists already
- conn.add_user adds the user to the mailadm.db
- mailcow.add_user adds the user to mailcow

Alternative 3 is my favorite I think, because the responsibilities are much more clear. But I think for now I would go with Alternative 1 and do Alt3 in the bot PR.